pkg/core/klinebuilder.go

```diff
@@ -0,0 +1,221 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/types"
+	"github.com/sirupsen/logrus"
+)
+
+// kLineBuilder is a kline builder for a symbol
+// It will accumulate trades and build klines
+//
+//go:generate callbackgen -type kLineBuilder
+type kLineBuilder struct {
+	symbol        string
+	minInterval   types.Interval
+	intervals     map[types.Interval]struct{}
+	isBackTesting bool
+
+	kLineCallbacks       []func(kline types.KLine)
+	kLineClosedCallbacks []func(kline types.KLine)
+	logger               *logrus.Entry
+
+	//
+	mu      sync.Mutex
+	state   *kLineBuilderState
+	running bool
+
+	exchange types.Exchange
+}
+
+// NewKLineBuilder: constructor of kLineBuilder
+//   - symbol: symbol to trace on
+//   - minInterval: unit interval, related to your signal timeframe.
+//     All the supported intervals of the binding stream should be multiple of this interval.
+func NewKLineBuilder(symbol string, minInterval types.Interval, isBackTesting bool) *kLineBuilder {
```

Comment:
you don't need to pass isBackTesting to this builder, we just need to make this component backtest-able by ensuring not using async in the process